cdap-cli/src/main/java/co/cask/cdap/cli/command/app/BaseBatchCommand.java

```diff
@@ -0,0 +1,142 @@
+import co.cask.cdap.cli.english.Fragment;
+import co.cask.cdap.cli.util.AbstractAuthCommand;
+import co.cask.cdap.client.ApplicationClient;
+import co.cask.cdap.common.ApplicationNotFoundException;
+import co.cask.cdap.common.UnauthorizedException;
+import co.cask.cdap.proto.BatchProgram;
+import co.cask.cdap.proto.Id;
+import co.cask.cdap.proto.ProgramRecord;
+import co.cask.cdap.proto.ProgramType;
+import co.cask.common.cli.Arguments;
+import com.google.common.base.Joiner;
+import com.google.common.base.Splitter;
+
+import java.io.IOException;
+import java.io.PrintStream;
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * Base class for commands that work on multiple programs in an application.
+ *
+ * @param <T> the type of input object for the batch request
+ */
+public abstract class BaseBatchCommand<T extends BatchProgram> extends AbstractAuthCommand {
+  private final ApplicationClient appClient;
+
+  protected BaseBatchCommand(ApplicationClient appClient, CLIConfig cliConfig) {
+    super(cliConfig);
+    this.appClient = appClient;
+  }
+
+  @Override
+  public void perform(Arguments arguments, PrintStream output) throws Exception {
+    Args<T> args = readArgs(arguments);
+    if (args.programs.isEmpty()) {
+      output.printf(String.format("application '%s' contains no programs of type '%s'",
+        args.appId.getId(), Joiner.on(',').join(args.programTypes)));
+      return;
+    }
+
+    runBatchCommand(output, args);
+  }
+
+  /**
+   * Reads arguments to get app id, program types, and list of input programs.
+   */
+  protected Args<T> readArgs(Arguments arguments)
+    throws ApplicationNotFoundException, UnauthorizedException, IOException {
+
+    String appName = arguments.get(ArgumentName.APP.getName());
+    Id.Application appId = Id.Application.from(cliConfig.getCurrentNamespace(), appName);
+
+    Set<ProgramType> programTypes = getDefaultProgramTypes();
+    if (arguments.hasArgument(ArgumentName.PROGRAM_TYPES.getName())) {
+      programTypes.clear();
+      String programTypesStr = arguments.get(ArgumentName.PROGRAM_TYPES.getName());
+      for (String programTypeStr : Splitter.on(',').trimResults().split(programTypesStr)) {
+        ProgramType programType = ProgramType.valueOf(programTypeStr.toUpperCase());
+        programTypes.add(programType);
+      }
+    }
+
+    List<T> programs = new ArrayList<>();
+    Map<ProgramType, List<ProgramRecord>> appPrograms = appClient.listProgramsByType(appId);
+    for (ProgramType programType : programTypes) {
+      List<ProgramRecord> programRecords = appPrograms.get(programType);
+      if (programRecords != null) {
+        for (ProgramRecord programRecord : programRecords) {
+          programs.add(createProgram(programRecord));
+        }
+      }
+    }
+    return new Args<>(appId, programTypes, programs);
+  }
+
+  protected Set<ProgramType> getDefaultProgramTypes() {
```

Comment:
the use case here is that somebody updates their application jar and needs to restart all their long running programs to pick up the new code.  So by default we are only including the long running program types (except for the status command, which overrides this method to include all types).

the reasoning here is that we only want to do long running program types, since we wouldn't want somebody to accidentally restart workflows and mapreduce jobs.